pylint/checkers/utils.py

```diff
@@ -1367,6 +1373,13 @@ def safe_infer(
             inferred_type = _get_python_type_of_node(inferred)
             if inferred_type not in inferred_types:
                 return None  # If there is ambiguity on the inferred node.
+            if (
```

Comment:
I'm not sure how the compiler handle this but since this is called so often we might want to "optimize" the if's here. If we use `if isinstance(inferred, ...)` and then go down a level we can use `elifs` to make sure we aren't doing `if` statement that we know will never be true.

Not big enough to create a follow-up but I guess something to consider for future PRs that touch this code.

Reply:
Good point. I'm not sure we'll be able to keep that in mind if we don't do the follow-up immediately though.

Reply:
> I'm not sure how the compiler handle this

`and` statements don't evaluate the right hand side if the left side tests false:
```python
>>> not 1 and (1 / 0)
False
```

Do you have a suggested optimization that would short-circuit more?

Reply:
Since it's merged I can't create an actual suggestion, but:
```python
isinstance(inferred, nodes.Const):
    if compare_constants ...
        ...
elif isinstance(inferred, nodes.FunctionDef):
    if inferred.args.args is not None:
      ...
```

Reply:
Ah. Perhaps. But we would need a microbenchmark for how many times we call `compare_constants`, as it's probably way faster to check the boolean, right?

Reply:
Yeah, you're right!